### DISCUSSION TITLE
How to inpaint an existing object on top of another existing object?

### DISCUSSION THREAD

**Discussion by levoz92** [huggingface/diffusers#8374] [General]
Hi there

![nike-shoe](https://github.com/huggingface/diffusers/assets/162665819/e1603c85-fa7f-4701-aa26-6d1731a53cff)

Above, I got this Nike shoe.

My goal is simple. Take a photo of an actor/actress wearing a pair of shoes and in paint this Nike instead of what they are wearing. How can I do this in a way where the inpainted nike's orientation/angle will match that of the shoe inside the picture?

**Comment by asomoza** [comment]
Are you asking from the perspective of someone that doesn't know how to do inpainting or are you a more advanced user that knows how to do good inpaints?

As a hobby or test project, you can just use IP Adapters to inpaint the shoes over a person.

In a professional level it's a lot longer answer:

To get the exact same level of details of the real object it's really hard, that's what all the `virtual try on (VTON)` solutions try to achieve. If you want a ready to go solution probably your best bet is to search and test them.

Here's a quick list but there's a lot more:

https://github.com/yisol/IDM-VTON
https://github.com/levihsu/OOTDiffusion
https://github.com/bcmi/DCI-VTON-Virtual-Try-On
https://github.com/xiezhy6/GP-VTON


The size of the object also affects, for shoes you could probably get good results with closeups but for a person standing with a full body shot, the results will be very bad, same as faces.

Sneakers are specially hard since the are affected by a common problem in diffusion models which is that they have a lot of tiny details that gets generated in a wrong way, same problem as fingers in hands, details in cars or bikes, etc.

For my experience, this is a lot of work if you want a professional solution, you'll need to train a LoRA, use IP adapters, controlnet and do a lot of refining after with inpainting to fix each detail to just get one production ready image, at this point I think it's still more time and cost efficient to do it with the traditional method of just hiring the model and taking the photos.

**Reply by levoz92** [reply]
My goal is simpler actually. Maybe I didn't articulate it well.

- if I got a shoe like this, how can I generate the legs/feet of a human to wear it?

**Reply by asomoza** [reply]
That's a lot different than the first question, also I'm still not clear of the quality level you want. Are you happy with something like this?

||||
|---|---|---|
|![20240531212240](https://github.com/huggingface/diffusers/assets/5442875/9aeede1e-2c98-47c0-bbb0-afb9be7f6f85)|![20240531213342](https://github.com/huggingface/diffusers/assets/5442875/a53f32b7-f592-4f8a-8d3a-8508adf489e4)|![20240531211745](https://github.com/huggingface/diffusers/assets/5442875/52d91406-3461-4679-a436-e37e5b760db5)|

**Reply by levoz92** [reply]
Yes that's great.

**Reply by asomoza** [reply]
for this I didn't use inpaint, I used controlnet with IP Adapters.

I removed the background from the shoe, and made it smaller and rotate it. Then I used anyline, which is a new controlnet that grabs a lot of the details, so I got this:

![20240531213627](https://github.com/huggingface/diffusers/assets/5442875/4fd4a9f1-f5e0-4ec3-8fcb-43dfd6fb81d3)


Then I used an IP Adapter Plus with the mask of the shoe and with a couple of style layers enabled to grab the colors and a little more of the details.

Then I used the prompt "hiqh quality photo of a person legs wearing sneakers", it generates the leg or legs half of the time though, so you need to keep generating until you get a good result.

I prefer this way because it looks a lot more realistic and seamless than inpainting.

If I find or maybe you can provide a more generic and with a free to use licence image of the shoes I can do a full example of this. All of the shoes I find are branded.

**Reply by levoz92** [reply]
Is there a way to do this with Huggingface as opposed to UIs like A1111?

**Reply by asomoza** [reply]
yeah, that's what I'm using. I think this is a good example for a guide, but to give you an example fast I'll just write the code.

I found some good generic shoes with a free licence, I'll post the code here as soon as I have it and test it.

**Reply by levoz92** [reply]
Will you post the code here?

**Reply by asomoza** [reply]
yes

**Reply by levoz92** [reply]
Thank you very much. Looking forward to it.

**Reply by levoz92** [reply]
Any updates?

**Reply by asomoza** [reply]
done, posted as another thread to separate it from this one.

**Comment by asomoza** [comment]
I'm going to use this photo from  unsplash:

<p align="center">
<img width="400" src="https://github.com/huggingface/diffusers/assets/5442875/53006aaf-7207-49fd-98a5-01cbfbc59c77">
</p>

Photo by <a href="https://unsplash.com/@pl_gt?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash">[Paul Gaudriault](https://unsplash.com/@pl_gt?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash)</a> on <a href="https://unsplash.com/photos/unpaired-maroon-plimsoll-on-top-of-yellow-textile-a-QH9MAAVNI?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash">[Unsplash](https://unsplash.com/photos/unpaired-maroon-plimsoll-on-top-of-yellow-textile-a-QH9MAAVNI?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash)</a>

For this part I will do it manually since it's faster and easier, if I do a guide about this I'll do this with code. So what we need is to rotate, crop and resize the image (better to work with 1024x1024 images with SDXL). After that we can just use code, so first is to pass the image with the anyline preprocessor and then generate the image with the mistoline controlnet and one IP adapter.

|source|anyline|result|
|---|---|---|
|![shoe_cropped_nobgpng_rotated_scaled](https://github.com/huggingface/diffusers/assets/5442875/2ee884bc-cd3d-4eb7-8da8-96abe94de5ad)|![anyline](https://github.com/huggingface/diffusers/assets/5442875/1afb4275-c241-4a42-bc46-a894f6ce0775)|![result](https://github.com/huggingface/diffusers/assets/5442875/c6ef84f2-5913-4447-abed-b418e8dddacd)|

Edit: added compel prompt weighting and changed a bit the params to get more consistent results.

Install first `controlnet_aux` and `compel`. Here's the code:

```python
import torch
from compel import Compel, ReturnedEmbeddingsType
from controlnet_aux import AnylineDetector

from diffusers import (
    AutoencoderKL,
    ControlNetModel,
    DPMSolverMultistepScheduler,
    StableDiffusionXLControlNetPipeline,
)
from diffusers.utils import load_image


source_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/shoe_cropped_nobgpng_rotated_scaled.png?download=true"
)
anyline = AnylineDetector.from_pretrained("TheMistoAI/MistoLine", filename="MTEED.pth", subfolder="Anyline").to("cuda")
preprocessed = anyline(source_image, detect_resolution=1024)

controlnet = ControlNetModel.from_pretrained(
    "TheMistoAI/MistoLine",
    torch_dtype=torch.float16,
    variant="fp16",
)

vae = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16).to("cuda")

pipeline = StableDiffusionXLControlNetPipeline.from_pretrained(
    "SG161222/RealVisXL_V4.0", torch_dtype=torch.float16, variant="fp16", controlnet=controlnet, vae=vae
).to("cuda")
pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config, use_karras_sigmas=True)

pipeline.load_ip_adapter(
    "h94/IP-Adapter",
    subfolder="sdxl_models",
    weight_name="ip-adapter-plus_sdxl_vit-h.safetensors",
    image_encoder_folder="models/image_encoder",
)

scale_config = {
    "up": {"block_0": [0.0, 0.9, 0.0]},
}
pipeline.set_ip_adapter_scale(scale_config)

compel = Compel(
    tokenizer=[pipeline.tokenizer, pipeline.tokenizer_2],
    text_encoder=[pipeline.text_encoder, pipeline.text_encoder_2],
    returned_embeddings_type=ReturnedEmbeddingsType.PENULTIMATE_HIDDEN_STATES_NON_NORMALIZED,
    requires_pooled=[False, True],
)
prompt = "hiqh quality photo of a (person leg)1.5 wearing a shoe+ on the sidewalk"
conditioning, pooled = compel(prompt)

generator = torch.Generator(device="cpu").manual_seed(216797721)

image = pipeline(
    prompt_embeds=conditioning,
    pooled_prompt_embeds=pooled,
    negative_prompt="bokeh",
    guidance_scale=7.5,
    num_inference_steps=25,
    generator=generator,
    image=preprocessed,
    controlnet_conditioning_scale=0.8,
    control_guidance_end=0.5,
    ip_adapter_image=source_image,
).images[0]

image.save("result.png")
```

**Reply by levoz92** [reply]
Why does this only produce one type of image? When I change the prompt to have the person running or have the background be bright daylight, it doesn't make a difference.

**Reply by asomoza** [reply]
I can only guide you and can't do all the use cases you want because it takes time, that's why I was asking about this at the beginning, where you changed your question to:

> if I got a shoe like this, how can I generate the legs/feet of a human to wear it?

So take this as the initial step and start experimenting with it, that's the only way you will learn how to do more advanced stuff.

Specifically to answer your new questions:

> have the background be bright daylight

I'm using the same image for the IP Adapter, since it's a transparent image it's taking the background as black, which makes the generation a more "night" setting, to change this you can use a mask over the IP Adapter image and to generate different background you can use a second IP Adapter with the background you want, with this you don't need a prompt to change the background or the time of the day.

> have the person running

The model can't do miracles, if there's only one shoe it will be really hard for it to generate a person running, also if they're both flat on the ground it won't be able to do it too or it will generate a very deformed person. So to generate a person running you'll need to duplicate the shoe and position them in a running position.

This all can be done with code, for example, save a preset of positions to put the shoes, or make legs that help the model. 

Also since you're changing a lot of the initial question, if what you want gets even more complex, at some point it's better to switch to an inpainting or image to image solution.

**Reply by levoz92** [reply]
Since the input is a shoe with the bg removed, isn't it possible to rotate it and feed it to the model to account for different oritentations?

**Reply by asomoza** [reply]
yeah exactly, you can do whatever you want with the shoe to help the model or to generate different positions. That's what I did when I was asking in the first post:

![image](https://github.com/huggingface/diffusers/assets/5442875/138e5ae3-af57-4f1b-a945-60e385661fab)

**Reply by levoz92** [reply]
Is it possible to create orientations of a shoe based on the angle of another one?
If I create a segmentation model that detects someone wearing shoes, how can we use your above logic to align the desired shoe to the same angle as the detected one?

**Reply by levoz92** [reply]
Also, my implementation ends up changing the design of the shoe. Why could that be?

**Reply by asomoza** [reply]
if you want the shoe to be more precise you need to play with the controlnet guidance params:

```python
    controlnet_conditioning_scale=0.8,
    control_guidance_end=0.5,
```
but If you make it too strong the generations will be bad because you're taking away the freedom of the model to adapt the generation.

> Is it possible to create orientations of a shoe based on the angle of another one?

Yes but that is not a `diffusers` question, that's just normal code. Sadly even if it interest me I don't have the time to also answer implementation questions. I suggest to search for this or even asking a chatbot on how to do it.

> If I create a segmentation model that detects someone wearing shoes, how can we use your above logic to align the desired shoe to the same angle as the detected one?

This is an interesting question, the segmentation models gives you an area and not an angle so no right out of the box. I don't really know if there's a model that gives you the angle too, but it should be possible to train one for this or even maybe code something that detects the angle of an object.

Your question is really more complex now and it's more like the first one, if you really want to preserve the full detail of the shoes it will be better to use an inpaint model with a trained LoRA or maybe do an image to image  (with low denoise) over the pasted shoes. 

Also you can try openpose or dwpose to detect the person and maybe try to guess the position of the shoes or the angle.

But these are all questions that are out of the scope of `diffusers` and are more full solutions like the virtual try on ones.
